extensions/gitpod/src/remoteConnector.ts

```diff
@@ -379,64 +380,186 @@ export default class LocalApp extends Disposable {
 		}
 	}
 
-	public async handleUri(uri: vscode.Uri) {
-		if (uri.path === LocalApp.AUTH_COMPLETE_PATH) {
-			this.logger.info('auth completed');
-			return;
+	private async getWorkspaceSSHDestination(workspaceId: string, gitpodHost: string): Promise<string> {
+		const session = await vscode.authentication.getSession(
+			'gitpod',
+			['function:getWorkspace', 'function:getOwnerToken', 'function:getLoggedInUser', 'resource:default'],
+			{ createIfNone: true }
+		);
+
+		const serviceUrl = new URL(gitpodHost);
+
+		const workspaceInfo = await withServerApi(session.accessToken, serviceUrl.toString(), service => service.server.getWorkspace(workspaceId), this.logger);
+		if (!workspaceInfo.latestInstance) {
+			this.logger.error('No running instance for this workspaceId', workspaceId);
+			throw new Error('no_running_instance');
 		}
-		this.logger.info('open workspace window: ' + uri.toString());
-		const params: SSHConnectionParams = JSON.parse(uri.query);
-		let resolvedConfig: LocalAppConfig | undefined;
-		try {
-			await vscode.window.withProgress({
-				location: vscode.ProgressLocation.Notification,
-				cancellable: true,
-				title: `Connecting to Gitpod workspace: ${params.workspaceId}`
-			}, async (_, token) => {
-				const connection = await this.withLocalApp(params.gitpodHost, (client, config) => {
-					resolvedConfig = config;
+
+		const workspaceUrl = new URL(workspaceInfo.latestInstance.ideUrl);
+
+		const sshHostKeyEndPoint = `https://${workspaceUrl.host}/_ssh/host_keys`;
+		const sshHostKeyResponse = await fetch(sshHostKeyEndPoint);
+		if (!sshHostKeyResponse.ok) {
+			// Gitpod SSH gateway not configured
+			this.logger.error('SSH gateway not configured for this Gitpod Host', gitpodHost);
+			throw new Error('no_ssh_gateway');
+		}
+
+		const ownerToken = await withServerApi(session.accessToken, serviceUrl.toString(), service => service.server.getOwnerToken(workspaceId), this.logger);
+
+		const sshDestInfo = {
+			user: `${workspaceId}#${ownerToken}`,
+			hostName: workspaceUrl.host
```

Comment:
It turend out that for self hosted we need a dedicated ssh subdomain, otherwise some customers cannot expose it via load balancers.

Please replace `workspaceId` part of host with `${workspaceId}.ssh`.